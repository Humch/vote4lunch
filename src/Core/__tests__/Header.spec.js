/* eslint-disable no-undef */
import React from 'react';
import reduxThunk from 'redux-thunk';
import { Provider } from 'react-redux';

import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';

import { BrowserRouter as Router } from 'react-router-dom';

// mock browser local storage
import LS from '../../../__mocks__/localStorage';

import { Header } from '..';


describe('Header Snapshot, screen larger than 768px', () => {
  const initialState = {
    voteData: {
      id: 0,
      pseudo: ''
    },
    topAlert: {
      messageType: 'success',
      status: true,
      message: 'Ceci est un test'
    }
  };
  const mockStore = configureStore();
  let store;

  beforeEach(() => {
    store = mockStore(initialState);
  });

  afterEach(() => {
    LS.window.localStorage.clear();
  });

  it('capturing Snapshot of Header, user is not logged in, screen larger than 768px', () => {
    const renderedValue = renderer.create(
      <Provider store={store}>
        <Header />
      </Provider>
    ).toJSON();
    expect(renderedValue).toMatchSnapshot();
  });

  it('capturing Snapshot of Header with pseudo and email, screen larger than 768px', () => {
    LS.window.localStorage.setItem('pseudo', 'bob');
    LS.window.localStorage.setItem('email', 'bob@bob.com');

    const div = document.createElement('div');
    div.setAttribute('id', 'TooltipMyVotes');
    document.body.appendChild(div);

    const renderedValue = renderer.create(
      <Router>
        <Provider store={store}>
          <Header />
        </Provider>
      </Router>
    ).toJSON();
    expect(renderedValue).toMatchSnapshot();
  });
});

describe('Header Snapshot, screen smaller than 768px', () => {
  const initialState = {
    voteData: {
      id: 0,
      pseudo: ''
    },
    topAlert: {
      messageType: 'success',
      status: true,
      message: 'Ceci est un test'
    }
  };
  const mockStore = configureStore();
  let store;

  beforeEach(() => {
    store = mockStore(initialState);
    window.innerWidth = 767;
  });

  afterEach(() => {
    LS.window.localStorage.clear();
  });

  it('capturing Snapshot of Header, user is not logged in, screen smaller than 768px', () => {
    const renderedValue = renderer.create(
      <Router>
        <Provider store={store}>
          <Header />
        </Provider>
      </Router>
    ).toJSON();
    expect(renderedValue).toMatchSnapshot();
  });

  it('capturing Snapshot of Header with pseudo and email, screen smaller than 768px', () => {
    LS.window.localStorage.setItem('pseudo', 'bob');
    LS.window.localStorage.setItem('email', 'bob@bob.com');

    const div = document.createElement('div');
    div.setAttribute('id', 'TooltipMyVotes');
    document.body.appendChild(div);

    const renderedValue = renderer.create(
      <Router>
        <Provider store={store}>
          <Header />
        </Provider>
      </Router>
    ).toJSON();
    expect(renderedValue).toMatchSnapshot();
  });
});

describe('Header (Shallow + passing the {store} directly)', () => {
  const initialState = {
    voteData: {
      id: 0,
      pseudo: ''
    }
  };
  const mockStore = configureStore();
  let store;
  let container;

  beforeEach(() => {
    store = mockStore(initialState);
    container = shallow(<Header store={store} />);
  });

  it('render the connected(SMART) component', () => {
    expect(container.length).toEqual(1);
  });

  it('check Prop matches with initialState', () => {
    expect(container.prop('voteData')).toEqual(initialState.voteData);
  });
});

describe('Header - REACT-REDUX (Mount + wrapping in <Provider>) - user is logged in', () => {
  const initialState = {
    voteData: {
      id: 0,
      pseudo: ''
    },
    topAlert: {
      messageType: 'success',
      status: true,
      message: 'Ceci est un test'
    }
  };
  const middlewares = [reduxThunk];
  const mockStore = configureStore(middlewares);
  let store;
  let wrapper;

  beforeEach(() => {
    store = mockStore(initialState);
    store.clearActions();

    // set localStorage value to login user
    LS.window.localStorage.setItem('pseudo', 'bob');
    LS.window.localStorage.setItem('email', 'bob@bob.com');

    wrapper = mount(
      <Router>
        <Provider store={store}>
          <Header />
        </Provider>
      </Router>
    );

    const div = document.createElement('div');
    div.setAttribute('id', 'TooltipMyVotes');
    document.body.appendChild(div);
  });

  afterEach(() => {
    // clear localStorage
    LS.window.localStorage.clear();
  });

  it('render the connected(SMART) component', () => {
    expect(wrapper.length).toEqual(1);
  });

  it('check if Hamburger is open', () => {
    window.innerWidth = 767;

    const header = wrapper.find('Header');
    expect(header.state().isOpen).toBeFalsy();
    const buttonHamburger = wrapper.find('.navbar-toggler');
    buttonHamburger.simulate('click');
    expect(header.state().isOpen).toBeTruthy();
  });

  it('check if ToolTip is open', () => {
    const header = wrapper.find('Header');
    const toolTip = wrapper.find('Tooltip');

    expect(header.state().tooltipOpen).toBeFalsy();
    toolTip.props().toggle();
    expect(header.state().tooltipOpen).toBeTruthy();
  });

  it('check if LoginModal is closed after login', () => {
    const header = wrapper.find('Header');
    const buttonLogout = wrapper.find('Button#logoutButton');
    header.state().openLoginModal = true;

    buttonLogout.simulate('click');
    expect(header.state().openLoginModal).toBeFalsy();
  });

  it('check if LoginModal is closed after login for screen smaller than 768px', () => {
    const header = wrapper.find('Header');
    const buttonLogout = wrapper.find('Button#logoutButton');
    header.state().openLoginModal = true;

    buttonLogout.simulate('click');
    expect(header.state().openLoginModal).toBeFalsy();
  });
});

describe('Header - REACT-REDUX (Mount + wrapping in <Provider>) - user is logged in', () => {
  const initialState = {
    voteData: {
      id: 0,
      pseudo: ''
    },
    topAlert: {
      messageType: 'success',
      status: true,
      message: 'Ceci est un test'
    }
  };
  const middlewares = [reduxThunk];
  const mockStore = configureStore(middlewares);
  let store;
  let wrapper;

  beforeEach(() => {
    store = mockStore(initialState);
    store.clearActions();

    window.innerWidth = 1280;

    // set localStorage value to login user
    LS.window.localStorage.setItem('pseudo', 'bob');
    LS.window.localStorage.setItem('email', 'bob@bob.com');

    wrapper = mount(
      <Router>
        <Provider store={store}>
          <Header />
        </Provider>
      </Router>
    );

    const div = document.createElement('div');
    div.setAttribute('id', 'TooltipMyVotes');
    document.body.appendChild(div);
  });

  afterEach(() => {
    // clear localStorage
    LS.window.localStorage.clear();
  });

  it('render the connected(SMART) component', () => {
    expect(wrapper.length).toEqual(1);
  });

  it('check if LoginModal is closed after login for screen larger than 768px', () => {
    const header = wrapper.find('Header');
    const buttonLogout = wrapper.find('Button#logoutButton');
    header.state().openLoginModal = true;

    buttonLogout.simulate('click');
    expect(header.state().openLoginModal).toBeFalsy();
  });
});

describe('Header - REACT-REDUX (Mount + wrapping in <Provider>) - user is not logged in', () => {
  const initialState = {
    voteData: {
      id: 0,
      pseudo: ''
    },
    topAlert: {
      messageType: 'success',
      status: true,
      message: 'Ceci est un test'
    },
    userData: {
      pseudo: '',
      email: '',
    },
    voteDataForm: {
      pseudo: '',
      email: '',
    }
  };
  const middlewares = [reduxThunk];
  const mockStore = configureStore(middlewares);
  let store;
  let wrapper;

  beforeEach(() => {
    store = mockStore(initialState);
    store.clearActions();

    wrapper = mount(
      <Router>
        <Provider store={store}>
          <Header />
        </Provider>
      </Router>
    );

    const div = document.createElement('div');
    div.setAttribute('id', 'TooltipMyVotes');
    document.body.appendChild(div);
  });

  afterEach(() => {
    LS.window.localStorage.clear();
  });

  it('render the connected(SMART) component', () => {
    expect(wrapper.length).toEqual(1);
  });

  it('check if Hamburger is open', () => {
    window.innerWidth = 767;

    const header = wrapper.find('Header');
    expect(header.state().isOpen).toBeFalsy();
    const buttonHamburger = wrapper.find('.navbar-toggler');
    buttonHamburger.simulate('click');
    expect(header.state().isOpen).toBeTruthy();
  });

  it('check if LoginModal is open', () => {
    const header = wrapper.find('Header');
    const buttonLogin = wrapper.find('Button#loginButton');

    expect(header.state().openLoginModal).toBeFalsy();
    buttonLogin.simulate('click');
    expect(header.state().openLoginModal).toBeTruthy();
  });
});
