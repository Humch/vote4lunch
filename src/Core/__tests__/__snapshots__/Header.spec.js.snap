// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Snapshot, screen larger than 768px capturing Snapshot of Header with pseudo and email, screen larger than 768px 1`] = `
<div
  className="Header"
>
  <div
    className="AlertBox alert alert-success fade"
    role="alert"
  >
    
    Ceci est un test
    
  </div>
  <nav
    className="bg-blue navbar navbar-expand-md navbar-dark"
  >
    <a
      className="navbar-brand"
      href="/"
    >
      Vote 4 Lunch
    </a>
    <button
      className="navbar-toggler"
      onClick={[Function]}
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      className="collapse navbar-collapse"
      style={Object {}}
    >
      <ul
        className="ml-auto navbar-nav"
      >
        <li
          className="mr-3 pt-1 nav-item"
        >
          <a
            className="nav-link"
            onClick={[Function]}
          >
            Bonjour, 
            bob
          </a>
        </li>
        <li
          className="mr-3 nav-item"
        >
          <a
            aria-label={null}
            className="btn btn-info"
            href="/my-votes"
            id="TooltipMyVotes"
            onClick={[Function]}
          >
            <i
              className="material-icons md-18 md-dark "
              icon="how_to_vote"
              style={
                Object {
                  "color": "white",
                  "fontSize": "md-18",
                }
              }
            />
          </a>
        </li>
        <li
          className="mr-3 nav-item"
        >
          <a
            aria-label={null}
            className="btn btn-danger"
            href="/logout"
            id="logoutButton"
            onClick={[Function]}
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-power-off fa-w-16 "
              data-icon="power-off"
              data-prefix="fas"
              role="img"
              style={Object {}}
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M400 54.1c63 45 104 118.6 104 201.9 0 136.8-110.8 247.7-247.5 248C120 504.3 8.2 393 8 256.4 7.9 173.1 48.9 99.3 111.8 54.2c11.7-8.3 28-4.8 35 7.7L162.6 90c5.9 10.5 3.1 23.8-6.6 31-41.5 30.8-68 79.6-68 134.9-.1 92.3 74.5 168.1 168 168.1 91.6 0 168.6-74.2 168-169.1-.3-51.8-24.7-101.8-68.1-134-9.7-7.2-12.4-20.5-6.5-30.9l15.8-28.1c7-12.4 23.2-16.1 34.8-7.8zM296 264V24c0-13.3-10.7-24-24-24h-32c-13.3 0-24 10.7-24 24v240c0 13.3 10.7 24 24 24h32c13.3 0 24-10.7 24-24z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </a>
        </li>
      </ul>
    </div>
  </nav>
  
</div>
`;

exports[`Header Snapshot, screen larger than 768px capturing Snapshot of Header, user is not logged in, screen larger than 768px 1`] = `
<div
  className="Header"
>
  <div
    className="AlertBox alert alert-success fade"
    role="alert"
  >
    
    Ceci est un test
    
  </div>
  <nav
    className="bg-blue navbar navbar-expand-md navbar-dark"
  >
    <a
      className="navbar-brand"
      href="/"
    >
      Vote 4 Lunch
    </a>
    <button
      className="navbar-toggler"
      onClick={[Function]}
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      className="collapse navbar-collapse"
      style={Object {}}
    >
      <ul
        className="ml-auto navbar-nav"
      >
        <li
          className="mr-3 nav-item"
        >
          <button
            aria-label={null}
            className="btn btn-outline-success"
            id="loginButton"
            onClick={[Function]}
            type="button"
          >
            S'identifier
          </button>
        </li>
        
        
      </ul>
    </div>
  </nav>
  
  
</div>
`;

exports[`Header Snapshot, screen smaller than 768px capturing Snapshot of Header with pseudo and email, screen smaller than 768px 1`] = `
<div
  className="Header"
>
  <div
    className="AlertBox alert alert-success fade"
    role="alert"
  >
    
    Ceci est un test
    
  </div>
  <nav
    className="bg-blue navbar navbar-expand-md navbar-dark"
  >
    <a
      className="navbar-brand"
      href="/"
    >
      Vote 4 Lunch
    </a>
    <button
      className="navbar-toggler"
      onClick={[Function]}
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      className="collapse navbar-collapse"
      style={Object {}}
    >
      <ul
        className="ml-auto navbar-nav"
      >
        <li
          className="mr-3 pt-1 nav-item"
        >
          <a
            className="nav-link"
            onClick={[Function]}
          >
            Bonjour, 
            bob
          </a>
        </li>
        <li
          className="mr-3 nav-item"
        >
          <div
            className="btn-group"
            role="group"
          >
            <a
              aria-label={null}
              className="btn btn-info"
              href="/my-votes"
              id="TooltipMyVotes"
              onClick={[Function]}
            >
              <i
                className="material-icons md-18 md-dark "
                icon="how_to_vote"
                style={
                  Object {
                    "color": "white",
                    "fontSize": "md-18",
                  }
                }
              />
            </a>
            <a
              aria-label={null}
              className="btn btn-danger"
              href="/logout"
              id="logoutButton"
              onClick={[Function]}
            >
              <svg
                aria-hidden="true"
                className="svg-inline--fa fa-power-off fa-w-16 "
                data-icon="power-off"
                data-prefix="fas"
                role="img"
                style={Object {}}
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M400 54.1c63 45 104 118.6 104 201.9 0 136.8-110.8 247.7-247.5 248C120 504.3 8.2 393 8 256.4 7.9 173.1 48.9 99.3 111.8 54.2c11.7-8.3 28-4.8 35 7.7L162.6 90c5.9 10.5 3.1 23.8-6.6 31-41.5 30.8-68 79.6-68 134.9-.1 92.3 74.5 168.1 168 168.1 91.6 0 168.6-74.2 168-169.1-.3-51.8-24.7-101.8-68.1-134-9.7-7.2-12.4-20.5-6.5-30.9l15.8-28.1c7-12.4 23.2-16.1 34.8-7.8zM296 264V24c0-13.3-10.7-24-24-24h-32c-13.3 0-24 10.7-24 24v240c0 13.3 10.7 24 24 24h32c13.3 0 24-10.7 24-24z"
                  fill="currentColor"
                  style={Object {}}
                />
              </svg>
            </a>
          </div>
        </li>
        
      </ul>
    </div>
  </nav>
  
</div>
`;

exports[`Header Snapshot, screen smaller than 768px capturing Snapshot of Header, user is not logged in, screen smaller than 768px 1`] = `
<div
  className="Header"
>
  <div
    className="AlertBox alert alert-success fade"
    role="alert"
  >
    
    Ceci est un test
    
  </div>
  <nav
    className="bg-blue navbar navbar-expand-md navbar-dark"
  >
    <a
      className="navbar-brand"
      href="/"
    >
      Vote 4 Lunch
    </a>
    <button
      className="navbar-toggler"
      onClick={[Function]}
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      className="collapse navbar-collapse"
      style={Object {}}
    >
      <ul
        className="ml-auto navbar-nav"
      >
        <li
          className="mr-3 nav-item"
        >
          <button
            aria-label={null}
            className="btn btn-outline-success"
            id="loginButton"
            onClick={[Function]}
            type="button"
          >
            S'identifier
          </button>
        </li>
        
        
      </ul>
    </div>
  </nav>
  
  
</div>
`;
